README.rst
========================
 REANA-Workflow-Commons
========================

.. image:: https://img.shields.io/travis/reanahub/reana-workflow-commons.svg
      :target: https://travis-ci.org/reanahub/reana-workflow-commons

.. image:: https://img.shields.io/coveralls/reanahub/reana-workflow-commons.svg
      :target: https://coveralls.io/r/reanahub/reana-workflow-commons

.. image:: https://readthedocs.org/projects/docs/badge/?version=latest
      :target: https://reana-workflow-commons.readthedocs.io/en/latest/?badge=latest

.. image:: https://badge.waffle.io/reanahub/reana.svg?label=Status%3A%20ready%20for%20work&title=Issues%20ready%20for%20work
      :target: https://waffle.io/reanahub/reana

.. image:: https://badges.gitter.im/Join%20Chat.svg
      :target: https://gitter.im/reanahub/reana?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge

.. image:: https://img.shields.io/github/license/reanahub/reana-workflow-commons.svg
      :target: https://github.com/reanahub/reana-workflow-commons/blob/master/COPYING

About
-----

REANA-Workflow-Commons is a component of the `REANA <http://www.reana.io/>`_ reusable and
reproducible research data analysis platform. It provides common utilities shared by the REANA runtime components.

Features
--------

- utility functions for runtime components

Useful links
------------

- `documentation <https://reana-workflow-commons.readthedocs.io/>`_
- `releases <https://github.com/reanahub/reana-workflow-commons/releases>`_
- `known issues <https://github.com/reanahub/reana-workflow-commons/issues>`_
- `source code <https://github.com/reanahub/reana-workflow-commons>`_

For more information about the REANA reusable analysis platform, please see the
`REANA project <http://www.reana.io>`_ home page and the general `REANA
documentation <http://reana.readthedocs.io/>`_ pages.


CHANGES.rst
Changes
=======

Version 0.3.0.dev20180703
--------------------------

- Initial public release.

.. admonition:: Please beware

   Please note that REANA is in an early alpha stage of its development. The
   developer preview releases are meant for early adopters and testers. Please
   don't rely on released versions for any production purposes yet.


run-tests.sh
#!/bin/sh
#
# This file is part of REANA.
# Copyright (C) 2018 CERN.
#
# REANA is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
# Suite 330, Boston, MA 02111-1307, USA.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization or
# submit itself to any jurisdiction.

#pydocstyle reana_workflow_commons && \
isort -rc -c -df **/*.py && \
check-manifest --ignore ".travis-*" && \
sphinx-build -qnNW docs docs/_build/html && \
python setup.py test && \
sphinx-build -qnNW -b doctest docs docs/_build/doctest


reana_workflow_commons/utils.py
# -*- coding: utf-8 -*-
#
# This file is part of REANA.
# Copyright (C) 2018 CERN.
#
# REANA is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
# Suite 330, Boston, MA 02111-1307, USA.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization or
# submit itself to any jurisdiction.

"""REANA Workflow Commons utils."""


reana_workflow_commons/config.py
# -*- coding: utf-8 -*-
#
# This file is part of REANA.
# Copyright (C) 2018 CERN.
#
# REANA is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
# Suite 330, Boston, MA 02111-1307, USA.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization or
# submit itself to any jurisdiction.

"""REANA Workflow Commons flask configuration."""

import os

SHARED_VOLUME_PATH = os.getenv('SHARED_VOLUME_PATH', '/reana/default')
"""Path to the mounted REANA shared volume."""

BROKER_URL = os.getenv('RABBIT_MQ_URL',
                       'message-broker.default.svc.cluster.local')

BROKER_USER = os.getenv('RABBIT_MQ_USER', 'test')

BROKER_PASS = os.getenv('RABBIT_MQ_PASS', '1234')

BROKER_PORT = os.getenv('RABBIT_MQ_PORT', 5672)

BROKER = os.getenv('RABBIT_MQ', 'amqp://{0}:{1}@{2}//'.format(BROKER_USER,
                                                              BROKER_PASS,
                                                              BROKER_URL))

STATUS_QUEUE = 'jobs-status'

EXCHANGE = ''

TYPE = ''

ROUTING_KEY = 'jobs-status'


reana_workflow_commons/version.py
# -*- coding: utf-8 -*-
#
# This file is part of REANA.
# Copyright (C) 2018 CERN.
#
# REANA is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
# Suite 330, Boston, MA 02111-1307, USA.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization or
# submit itself to any jurisdiction.

"""Version information for REANA-Workflow-Commons.

This file is imported by ``reana_workflow_commons.__init__``
and parsed by ``setup.py``.
"""

from __future__ import absolute_import, print_function

__version__ = "0.3.0.dev20180703"


reana_workflow_commons/__init__.py
# -*- coding: utf-8 -*-
#
# This file is part of REANA.
# Copyright (C) 2018 CERN.
#
# REANA is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
# Suite 330, Boston, MA 02111-1307, USA.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization or
# submit itself to any jurisdiction.

"""REANA-Workflow-Commons."""

from __future__ import absolute_import, print_function

from .version import __version__

__all__ = ('__version__', )


reana_workflow_commons/publisher.py
# -*- coding: utf-8 -*-
#
# This file is part of REANA.
# Copyright (C) 2018 CERN.
#
# REANA is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# REANA is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# REANA; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
# Suite 330, Boston, MA 02111-1307, USA.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization or
# submit itself to any jurisdiction.

import json
import logging

import pika

from .config import (BROKER_PASS, BROKER_PORT,
                     BROKER_URL, BROKER_USER,
                     EXCHANGE, ROUTING_KEY,
                     STATUS_QUEUE)


class Publisher:
    """Progress publisher to MQ."""

    def __init__(self):
        self.broker_credentials = pika.PlainCredentials(BROKER_USER,
                                                        BROKER_PASS)
        self._params = pika.connection.ConnectionParameters(
            BROKER_URL, BROKER_PORT, '/', self.broker_credentials)
        self._conn = None
        self._channel = None

    def connect(self):
        """Connect to MQ channel."""
        if not self._conn or self._conn.is_closed:
            self._conn = pika.BlockingConnection(self._params)
            self._channel = self._conn.channel()
            self._channel.queue_declare(queue=STATUS_QUEUE)

    def _publish(self, msg):
        self._channel.basic_publish(exchange=EXCHANGE,
                                    routing_key=ROUTING_KEY,
                                    body=json.dumps(msg))
        logging.debug('Publisher: message sent: %s', msg)

    def publish_workflow_status(self, workflow_uuid, status,
                                logs='', message=None):
        """Publish msg, reconnecting if necessary."""
        try:
            msg = {"workflow_uuid": workflow_uuid,
                   "logs": logs,
                   "status": status,
                   "message": message}
            self._publish(msg)
        except pika.exceptions.ConnectionClosed:
            logging.debug('Publisher: ConnectionClosed, reconnecting.')
            print('Publisher: ConnectionClosed, reconnecting.')
            self.connect()
            self._publish(msg)

    def close(self):
        if self._conn and self._conn.is_open:
            logging.debug('Publisher: closing queue connection')
            self._conn.close()
